Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+functions.  As such the use is expected to be limited, but performance critical.
+
+#### Intrinsic `IsSupported` granularity
+
+The existing implementation is using `IsSupported` to determine whether an existing
+namespace is supported on a given platform.
+
+The granularity is kept at a large scale to have large blocks of functionality controlled
+by a single `IsSupported`.
+
+#### Intrinsic granularity
+
+In general individual intrinsic will be chosen to be fine grained.  These will generally correspond to a single assembly instruction.
+
+#### Intrinsics & Crossgen
+
+For any intrinsic which may not be supported on all variants of a platform.  Crossgen
+Method compilation must be trapped, so that the JIT is forced to generate optimal platform dependent 
+code at runtime.
+
+#### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
+
+### Choice of Arm64 naming conventions
+
+#### Namespaces
+
+`ARM32` and `ARM64` will follow similar namespace conventions.
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.Arm` is used for enumerations shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Base` is recommended for functionality included in all `ARM32` and `ARM64` platforms
+  + Do we need this or could we assume `ARMv8` is minimum supported platform?
+  + Should we assume `ARMv7` or `ARMv5` is minimum supported platform?
++ `System.Runtime.Intrinsics.Arm.ARMv8` is used for `ARMv8` added functionality included in both `ARM32` and `ARM64` platforms 
++ `System.Runtime.Intrinsics.Arm.ARMv8.Arm64` is used for `ARMv8` functionality only included in `ARM64` platform
++ `System.Runtime.Intrinsics.Arm.ARMv8_1....` is used for `ARMv8.1` functionality.  
+  + Use similar sub namespace names for only included in `ARM64` platform ...
++ `System.Runtime.Intrinsics.Arm.XXX` is used for optional features `CRC32`, `AES`, `SHA1` ... 
+  + All extensions which were not required when they were initially introduced
+  + For example, `CRC32` became required with `ARMv8.1`, but since it was introduced as optional in `ARMv8` it would go in its own namespace
+  
+If an extension is only supported on `Arm32`, there is no reason to add a sub namespace.
+
++ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+
+If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+
++ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
+
+Namespace is chosen based on what could be implemented, not on what is or will be implemented.
+
+#### Names
+
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named assembly instructions.
+
+`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
```

Comment:
Is there a similar confusion about namespace vs. class/struct here and other places?

IsSupported is a property. It needs to live on a type. What is the proposed namespace.name of the type?

Reply:
> Is there a similar confusion about namespace vs. class/struct here and other places?

Yes.  It is all confused.  Since I think visually, when I translated to word, I mispoke every where.   

> IsSupported is a property. It needs to live on a type. What is the proposed namespace.name of the type?

I intended the proposal to imply that for the `ARMv8.1` extension, the class/struct `System.Runtime.Intrinsics.Arm.ARMv8_1` would contain `IsSupported`

Reply:
If we are following the existing conventions that the Intel intrinsic classes were named on, I believe this should be `Arm.Armv81` (`Arm` rather than `ARM` and no `_`). Ex: `X86.Sse41`

Reply:
> `Arm.Armv81`

+ Dropping a decimal point completely seems to be a bad general precedence as `1.1` and `11` would be indistinguishable.  I would reject X86 choice as precedent setting/forcing.
+ `Arm.ArmV8_1` would be more readable

Reply:
I don't disagree that `ARMv8_1` is more readable (I would have preferred we use `x86.SSE4_1` as well).

However, I do believe that the two namespaces should follow the same convention (and that may involved fixing up the names in the `x86` namespace). CC @terrajobst, @karelz